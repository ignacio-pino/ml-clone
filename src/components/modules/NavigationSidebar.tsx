import { Box, HStack, Heading, Skeleton, Stack, Text } from "@chakra-ui/react";
import { ICategoryData, IFilterDataList } from "../../Interfaces";
import CategoryList from "../common/CategoryList";
import CategoryBreadcrumbs from "../common/CategoryBreadcrumbs";
import FilterBadge from "../common/FilterBadge";

export default function NavigationSidebar({
  filterDataList,
  categoryData,
  isLoading,
}: {
  filterDataList: IFilterDataList;
  categoryData: ICategoryData;
  isLoading: boolean;
}) {
  return (
    <Stack ml={"10vw"} w={"20vw"}>
      <Skeleton isLoaded={!isLoading} h="100vh">
        {categoryData.categoryInformation?.path_from_root && (
          <CategoryBreadcrumbs
            pathFromRoot={categoryData.categoryInformation?.path_from_root}
            categoryId={categoryData.categoryInformation?.id}
          />
        )}
        <Heading>
          {categoryData.categoryInformation
            ? categoryData.categoryInformation.name
            : categoryData.query
            ? categoryData.query.charAt(0).toUpperCase() +
              categoryData.query.slice(1)
            : "Sin resultados"}
        </Heading>
        <Text>{categoryData.totalResults} resultados</Text>

        {filterDataList.filters && (
          <HStack>
            {" "}
            {filterDataList.filters.map(({ id, name, values }) => {
              return id === "category" ? null : (
                <FilterBadge id={id} name={name} values={values} type="" />
              );
            })}
          </HStack>
        )}

        <Box>
          {categoryData.childrenCategories && (
            <CategoryList
              id={"category"}
              name="CategorÃ­as"
              type="text"
              values={categoryData.childrenCategories}
            />
          )}
          {filterDataList.availableFilters.map(({ id, name, values, type }) => (
            <CategoryList id={id} name={name} values={values} type={type} />
          ))}
        </Box>
      </Skeleton>
    </Stack>
  );
}
