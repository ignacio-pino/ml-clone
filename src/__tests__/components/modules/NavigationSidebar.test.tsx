import { render, screen } from "@testing-library/react";
import NavigationSidebar from "../../../components/modules/NavigationSidebar";
import { MemoryRouter } from "react-router-dom";

const mockedUseNavigate = vitest.fn();

vitest.mock("react-router-dom", () => ({
  ...require("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("NavigationSidebar Component", () => {
  const mockCategoryData = {
    categoryInformation: {
      id: "12345",
      name: "TestCategory",
      path_from_root: [{ id: "Test path", name: "test name" }],
    },
    query: "",
    totalResults: 10,
    childrenCategories: [],
  };

  const mockFilterDataList = {
    filters: [
      {
        id: "category",
        name: "TestCategory",
        values: [{ id: "test", name: "test category name" }],
        type: "text",
      },
      {
        id: "filter1",
        name: "TestFilter1",
        values: [{ id: "test", name: "test filter 1 name" }],
        type: "text",
      },
    ],
    availableFilters: [
      {
        id: "availableFilter1",
        name: "TestAvailableFilter1",
        values: [{ id: "test", name: "test available filter 1 name" }],
        type: "text",
      },
    ],
  };

  it("matches snapshot", () => {
    const container = render(
      <MemoryRouter>
        <NavigationSidebar
          filterDataList={mockFilterDataList}
          categoryData={mockCategoryData}
        />
      </MemoryRouter>
    ).container;
    expect(container).toMatchSnapshot();
  });

  it("renders the category name if categoryInformation is present", () => {
    render(
      <MemoryRouter>
        <NavigationSidebar
          filterDataList={mockFilterDataList}
          categoryData={mockCategoryData}
        />
      </MemoryRouter>
    );
    const categoryName = screen.getByText("TestCategory");
    expect(categoryName).toBeInTheDocument();
  });

  it("renders capitalized query if query is present and categoryInformation is not", () => {
    const alteredMockCategoryData = {
      ...mockCategoryData,
      categoryInformation: undefined,
      query: "testquery",
    };
    render(
      <MemoryRouter>
        <NavigationSidebar
          filterDataList={mockFilterDataList}
          categoryData={alteredMockCategoryData}
        />
      </MemoryRouter>
    );
    const queryName = screen.getByText("Testquery");
    expect(queryName).toBeInTheDocument();
  });

  it("renders 'Sin resultados' if neither categoryInformation nor query is present", () => {
    const alteredMockCategoryData = {
      ...mockCategoryData,
      categoryInformation: undefined,
      query: "",
    };
    render(
      <MemoryRouter>
        <NavigationSidebar
          filterDataList={mockFilterDataList}
          categoryData={alteredMockCategoryData}
        />
      </MemoryRouter>
    );
    const noResults = screen.getByText("Sin resultados");
    expect(noResults).toBeInTheDocument();
  });

  it("renders total results correctly", () => {
    render(
      <MemoryRouter>
        <NavigationSidebar
          filterDataList={mockFilterDataList}
          categoryData={mockCategoryData}
        />
      </MemoryRouter>
    );
    const totalResults = screen.getByText("10 resultados");
    expect(totalResults).toBeInTheDocument();
  });

  it("renders FilterBadge for each filter except 'category'", () => {
    render(
      <MemoryRouter>
        <NavigationSidebar
          filterDataList={mockFilterDataList}
          categoryData={mockCategoryData}
        />
      </MemoryRouter>
    );
    expect(screen.queryByText("test category name")).not.toBeInTheDocument();
    expect(screen.getByText("test filter 1 name")).toBeInTheDocument();
  });

  it("renders CategoryList for childrenCategories", () => {
    const alteredMockCategoryData = {
      ...mockCategoryData,
      childrenCategories: [
        { id: "testchildren1", name: "Child Category 1", results: 10 },
      ],
    };
    render(
      <MemoryRouter>
        <NavigationSidebar
          filterDataList={mockFilterDataList}
          categoryData={alteredMockCategoryData}
        />
      </MemoryRouter>
    );
    const childCategory = screen.getByText("Child Category 1");
    expect(childCategory).toBeInTheDocument();
  });

  it("renders CategoryList for each availableFilter", () => {
    render(
      <MemoryRouter>
        <NavigationSidebar
          filterDataList={mockFilterDataList}
          categoryData={mockCategoryData}
        />
      </MemoryRouter>
    );
    const availableFilter = screen.getByText("TestAvailableFilter1");
    expect(availableFilter).toBeInTheDocument();
  });
});
