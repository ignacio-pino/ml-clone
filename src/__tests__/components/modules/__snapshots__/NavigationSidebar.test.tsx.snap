// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationSidebar Component > matches snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-1t05was"
  >
    <nav
      aria-label="breadcrumb"
      class="chakra-breadcrumb css-1rsh7lu"
    >
      <ol
        class="chakra-breadcrumb__list css-70qvj9"
      >
        <li
          class="chakra-breadcrumb__list-item css-18biwo"
        >
          <a
            class="chakra-breadcrumb__link css-0"
            id="Test path"
          >
             
            test name
          </a>
        </li>
      </ol>
    </nav>
    <h2
      class="chakra-heading css-0"
    >
      TestCategory
    </h2>
    <p
      class="chakra-text css-0"
    >
      10
       resultados
    </p>
    <div
      class="chakra-stack css-1igwmid"
    >
      <span
        class="css-12nbfm5"
      >
        <span
          class="css-zvlevn"
        >
          test filter 1 name
        </span>
        <button
          aria-label="Eliminar filtro"
          class="css-1t3z1vh"
          type="button"
        >
          <svg
            class="chakra-icon css-1md6f3y"
            focusable="false"
            viewBox="0 0 512 512"
          >
            <path
              d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
              fill="currentColor"
            />
          </svg>
        </button>
      </span>
    </div>
    <div
      class="css-0"
    >
      <div
        class="chakra-stack css-8g8ihq"
        id="category"
      >
        <div
          class="css-s98few"
        >
          <b
            class="chakra-text css-0"
          >
            Categor√≠as
          </b>
        </div>
      </div>
      <div
        class="chakra-stack css-8g8ihq"
        id="availableFilter1"
      >
        <div
          class="css-s98few"
        >
          <b
            class="chakra-text css-0"
          >
            TestAvailableFilter1
          </b>
        </div>
      </div>
    </div>
  </div>
</div>
`;
